File: Algorithm/08、希尔排序/shell.go
package main

import "fmt"

func main() {
	array := []int{5, 9, 8, 7, 6, 4, 3, 2, 1}
	shell(array)
	fmt.Println(array)
}

func shell(array []int) {
	length := len(array)
	gap := length / 2
	for gap > 0 {

		for i := gap; i < length; i++ {

			for j := i; j-gap >= 0; j -= gap {
				if array[j] < array[j-gap] {
					array[j], array[j-gap] = array[j-gap], array[j]
				}
			}
		}

		gap /= 2
	}

}

File: Algorithm/08、希尔排序/shell.php
<?php

$a =[0, 9, 1, 5, 8, 3, 7, 6, 4, 2];
shell($a);

function shell($arr)
{
    $len = count($arr);
    $gap = floor($len / 2);
    while ($gap > 0) {

        for ($i = $gap; $i < $len; $i++) {

            for ($j = $i; $j - $gap >= 0; $j -= $gap) {

                if ($arr[$j] < $arr[$j - $gap]) {
                    $temp    = $arr[$j];
                    $arr[$j] = $arr[$j - $gap];
                    $arr[$j - $gap] = $temp;
                }
            }
        }

        $gap = floor($gap / 2);
    }
    print_r($arr);
}
